readthedocs/notifications/querysets.py

```diff
@@ -22,13 +24,63 @@ def add(self, *args, **kwargs):
             attached_to_content_type=content_type,
             attached_to_id=attached_to.id,
             message_id=message_id,
+            # Update only ``READ`` and ``UNREAD`` notifications because we want
+            # to keep track of ``DISMISSED`` and ``CANCELLED`` ones.
+            state__in=(UNREAD, READ),
         ).first()
 
         if notification:
             self.filter(pk=notification.pk).update(
-                *args, modified=timezone.now(), **kwargs
+                *args,
+                modified=timezone.now(),
+                state=UNREAD,
+                **kwargs,
             )
             notification.refresh_from_db()
             return notification
 
         return super().create(*args, attached_to=attached_to, **kwargs)
+
+    def cancel(self, message_id, attached_to):
+        """
+        Cancel an on-going notification because the underlying state has changed.
+
+        When a notification is not valid anymore because the user has made the
+        required action (e.g. paid an unpaid subscription) we use this method
+        to mark those notifications as ``CANCELLED``.
+
+        It only cancels notifications that are ``UNREAD`` or ``READ``.
+        """
+        content_type = ContentType.objects.get_for_model(attached_to)
+
+        self.filter(
+            attached_to_content_type=content_type,
+            attached_to_id=attached_to.id,
+            message_id=message_id,
+            state__in=(UNREAD, READ),
```

Comment:
This could get pretty slow if we have a huge Notifications table. I assume we have an index on some subset of these that hopefully makes it faster?

Reply:
We have an index on `state` that I think will work fine here. We should probably create an `index_together` on `attached_to_content_type` and `attached_to_id`, which is something we will use a lot: https://docs.djangoproject.com/en/5.0/ref/models/options/#index-together

Reply:
I went ahead and created a PR for this at https://github.com/readthedocs/readthedocs.org/pull/11050